src/ui/src/js/services/permission_service.js

```diff
@@ -0,0 +1,137 @@
+permissionService.$inject = ['$rootScope'];
+
+/**
+ * permissionService - Used for filtering logic based on User roles
+ * @param  {$rootScope} $rootScope Angular's $rootScope object
+ * @return {Object}                For use by a controller.
+ */
+export default function permissionService($rootScope) {
+    const service = {
+        hasPermission: (permission, global = false, garden_name = null, namespace = null, system_name = null, system_version = null, command_name = null, instance_name = null) => {
+            if (!$rootScope.config.authEnabled) return true;
+            if (_.isUndefined($rootScope.user)) return false;
+
+            let permissions = service.get_permissions(permission);
+
+            if ($rootScope.user.local_roles !== undefined) {
+                for (let i = 0; i < $rootScope.user.local_roles.length; i++) {
+                    if (permissions.includes($rootScope.user.local_roles[i].permission)) {
+                        if (service.checkRole($rootScope.user.local_roles[i], global = global, garden_name = garden_name, namespace = namespace, system_name = system_name, instance_name = instance_name, system_version = system_version, command_name = command_name)) {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            if ($rootScope.user.upstream_roles !== undefined) {
+                for (let i = 0; i < $rootScope.user.upstream_roles.length; i++) {
+                    if (permissions.includes($rootScope.user.upstream_roles[i].permission)) {
+                        if (service.checkRole($rootScope.user.upstream_roles[i], global = global, garden_name = garden_name, namespace = namespace, system_name = system_name, instance_name = instance_name, system_version = system_version, command_name = command_name)) {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            return false;
+        },
+        get_permissions: (permission) => {
```

Comment:
What about if someone overrides this with a local copy? Can they get access to permissions that they otherwise would not have?

Reply:
Yes and no, there are two layers of security, API and UI. When we release the UI it is compiled, so the source files are not available. So if they get permission to see a button or page, they won't be able to modify/see records they didn't originally have permissions to see.